### PR TITLE
BUG: fixed json_normalize for subrecords with NoneTypes (#20030)

### LINKED ISSUE #20030: json_normalize does not normalize subrecords properly if any subrecords values are NoneType
### Code Sample, a copy-pastable example if possible

```python
data_fail_to_normalize = \
        [{'info': None}, 
         
        {'info': 
         {'created_at': '11/08/1993', 'last_updated': '26/05/2012'},
        'author_name': 
         {'first': 'Jane', 'last_name': 'Doe'}
        }]

data_partial_fail = \
        [{'info': None, 
         'author_name': 
         {'first': 'Smith', 'last_name': 'Appleseed'}
        }, 
        
        {'info': 
         {'created_at': '11/08/1993', 'last_updated': '26/05/2012'},
        'author_name': 
         {'first': 'Jane', 'last_name': 'Doe'}
        }]

>>> import pandas as pd
>>> pd.io.json.json_normalize(data_fail_to_normalize)
```
#### Output 1
|    | author_name                           | info                                                       |
|---:|:--------------------------------------|:-----------------------------------------------------------|
|  0 | nan                                   |                          None                              |
|  1 | {'first': 'Jane', 'last_name': 'Doe'} | {'created_at': '11/08/1993', 'last_updated': '26/05/2012'} |

```python
>>> pd.io.json.json_normalize(data_partial_fail)
```
#### Output 2
|    | author_name.first   | author_name.last_name   |   info | info.created_at   | info.last_updated   |
|---:|:--------------------|:------------------------|-------:|:------------------|:--------------------|
|  0 | Smith               | Appleseed               |    nan | nan               | nan                 |
|  1 | Jane                | Doe                     |    nan | 11/08/1993        | 26/05/2012          |

### Problem description
I expected that the json_normalize function takes into account the presence of NoneTypes in the dictionaries. This leads to 2 separate issues (If I should open this as 2 separate issues, let me know). 

I have already written a fix that solves this issue - if anyone else can validate that this is not working as intended, I can set up a PR.

#### Output 1
Does not unnest json after encountering NoneType at first instance of subrecord, line 192 of [`pandas/io/json/normalize.py`](https://github.com/pandas-dev/pandas/blob/a00154dcfe5057cb3fd86653172e74b6893e337d/pandas/io/json/normalize.py#L192)

#### Output 2
Keeps the None value when encountered, [{k: {'alpha': 'foo', 'beta': 'bar'}}, {k: None}], see [`nested_to_record`](https://github.com/pandas-dev/pandas/blob/a00154dcfe5057cb3fd86653172e74b6893e337d/pandas/io/json/normalize.py#L24) function. Creates additional column of `nan`s which would not otherwise occur if that particular key was removed.

#### Output 1
|    | author_name.first   | author_name.last_name   | info.created_at   | info.last_updated   |
|---:|:--------------------|:------------------------|:------------------|:--------------------|
|  0 | nan                 | nan                     | nan               | nan                 |
|  1 | Jane                | Doe                     | 11/08/1993        | 26/05/2012          |

#### Output 2
|    | author_name.first   | author_name.last_name   | info.created_at   | info.last_updated   |
|---:|:--------------------|:------------------------|:------------------|:--------------------|
|  0 | Smith               | Appleseed               | nan               | nan                 |
|  1 | Jane                | Doe                     | 11/08/1993        | 26/05/2012          |

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-104-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: None
pip: 9.0.1
setuptools: 38.4.0
Cython: None
numpy: 1.13.1
scipy: None
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by aerymilts**
TST: additional coverage for the test cases 

DOC: added changes to whatsnew/v0.23.0.txt

Checklist for other PRs (remove this part if you are doing a PR for the pandas documentation sprint):

- [x] closes #20030
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/json/test_normalize.py [hunk 1]
```diff
@@ -226,6 +237,21 @@ def test_non_ascii_key(self):
         result = json_normalize(json.loads(testjson))
         tm.assert_frame_equal(result, expected)
 
+    def test_missing_field(self, author_missing_data):
+        result = json_normalize(author_missing_data)
```
Comment (jreback) [inline pandas/tests/io/json/test_normalize.py]: can you add the issue number as a comment. & a 1-liner explaining the test purpose

File: pandas/tests/io/json/test_normalize.py [hunk 2]
```diff
@@ -322,3 +348,26 @@ def test_json_normalize_errors(self):
                             ['general', 'trade_version']],
                       errors='raise'
                       )
+
+    def test_nonetype_dropping(self):
+        data = [
```
Comment (jreback) [inline pandas/tests/io/json/test_normalize.py]: same, pls add the gh issue number as a comment and 1 liner

File: pandas/tests/io/json/test_normalize.py [hunk 3]
```diff
@@ -226,6 +237,21 @@ def test_non_ascii_key(self):
         result = json_normalize(json.loads(testjson))
         tm.assert_frame_equal(result, expected)
 
+    def test_missing_field(self, author_missing_data):
+        result = json_normalize(author_missing_data)
+        ex_data = [
+            {'author_name.first': float('nan'),
```
Comment (jreback) [inline pandas/tests/io/json/test_normalize.py]: use np.nan here

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. some minor comments. ping on green.

**Comment by aerymilts** [comment]
@jreback Should I squash the changes?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @aerymilts nice patch! keep em coming!

**Pull request merged**